File: app/src/main/java/jp/gr/java_conf/snake0394/loglook_android/view/fragment/DamagedShipFragment.java
package jp.gr.java_conf.snake0394.loglook_android.view.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;

import java.util.ArrayList;

import jp.gr.java_conf.snake0394.loglook_android.R;
import jp.gr.java_conf.snake0394.loglook_android.bean.MyShip;
import jp.gr.java_conf.snake0394.loglook_android.bean.MyShipManager;
import jp.gr.java_conf.snake0394.loglook_android.storage.DamagedShipFragmentPrefs;
import jp.gr.java_conf.snake0394.loglook_android.storage.DamagedShipFragmentPrefsSpotRepository;

public class DamagedShipFragment extends Fragment {

    private RecyclerView recyclerView;
    private DamagedShipFragmentPrefs prefs;

    public DamagedShipFragment() {
        // Required empty public constructor
    }

    public static DamagedShipFragment newInstance() {
        DamagedShipFragment fragment = new DamagedShipFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        this.prefs = DamagedShipFragmentPrefsSpotRepository.getEntity(getContext());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View rootView = inflater.inflate(R.layout.fragment_damaged_ship, container, false);
        if (recyclerView == null) {
            recyclerView = (RecyclerView) rootView.findViewById(R.id.recycler_view);
        }

        Spinner spinner = (Spinner) rootView.findViewById(R.id.sortSpinner);
        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                Spinner type = (Spinner) parent;
                // 初回起動時の動作
                if (type.isFocusable() == false) {
                    type.setFocusable(true);
                    return;
                }
                prefs.sortType = (String) type.getSelectedItem();
                DamagedShipAdapter adapter = new DamagedShipAdapter(getFragmentManager(), prefs.sortType, prefs.order);
                recyclerView.swapAdapter(adapter, false);
                initDataList();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
        ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item);
        adapter.add("修復時間");
        adapter.add("損傷度");
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);
        spinner.setFocusable(false);
        spinner.setSelection(adapter.getPosition(prefs.sortType));


        Button orderButton = (Button) rootView.findViewById(R.id.button_order);
        orderButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Button orderButton = (Button) v;
                String buttonText = String.valueOf(orderButton.getText());
                switch (buttonText) {
                    case "降順":
                        prefs.order = "昇順";
                        break;
                    case "昇順":
                        prefs.order = "降順";
                        break;
                }
                orderButton.setText(prefs.order);
                DamagedShipAdapter adapter = new DamagedShipAdapter(getFragmentManager(), prefs.sortType, prefs.order);
                recyclerView.swapAdapter(adapter, false);
                initDataList();
            }
        });
        orderButton.setText(this.prefs.order);

        return rootView;
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        if (recyclerView.getAdapter() == null) {
            DamagedShipAdapter adapter = new DamagedShipAdapter(getFragmentManager(), prefs.sortType, prefs.order);
            recyclerView.setAdapter(adapter);
            initDataList();
        }
    }

    private void initDataList() {
        ((DamagedShipAdapter) recyclerView.getAdapter()).clearData();
        ArrayList<MyShip> initialDataList = new ArrayList<>();
        for (MyShip myShip : MyShipManager.INSTANCE.getMyShips()) {
            if (myShip == null) {
                continue;
            }
            if (myShip.getNdockTime() > 0) {
                initialDataList.add(myShip);
            }
        }
        ((DamagedShipAdapter) recyclerView.getAdapter()).addDataOf(initialDataList);
        recyclerView.scrollToPosition(0);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        DamagedShipFragmentPrefsSpotRepository.putEntity(getContext(), this.prefs);
    }
}

File: app/src/main/java/jp/gr/java_conf/snake0394/loglook_android/view/fragment/DamagedShipFragment.kt
package jp.gr.java_conf.snake0394.loglook_android.view.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.Spinner
import io.realm.Realm
import jp.gr.java_conf.snake0394.loglook_android.R
import jp.gr.java_conf.snake0394.loglook_android.bean.MyShip
import jp.gr.java_conf.snake0394.loglook_android.storage.DamagedShipFragmentPrefs
import kotlinx.android.synthetic.main.fragment_damaged_ship.*

class DamagedShipFragment : Fragment() {

    private val prefs by lazy { DamagedShipFragmentPrefs(context) }
    private val realm by lazy { Realm.getDefaultInstance() }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater!!.inflate(R.layout.fragment_damaged_ship, container, false)

        val spinner = rootView.findViewById(R.id.sortSpinner) as Spinner
        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                val type = parent as Spinner
                // 初回起動時の動作
                if (type.isFocusable == false) {
                    type.isFocusable = true
                    return
                }
                prefs.sortType = type.selectedItem as String
                val adapter = DamagedShipAdapter(fragmentManager, prefs.sortType, prefs.order)
                recycler_view.swapAdapter(adapter, false)
                initDataList()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {}
        }
        val adapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_item)
        adapter.add("修復時間")
        adapter.add("損傷度")
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = adapter
        spinner.isFocusable = false
        spinner.setSelection(adapter.getPosition(prefs.sortType))


        val orderButton = rootView.findViewById(R.id.button_order) as Button
        orderButton.setOnClickListener { v ->
            val orderButton = v as Button
            val buttonText = orderButton.text.toString()
            when (buttonText) {
                "降順" -> prefs.order = "昇順"
                "昇順" -> prefs.order = "降順"
            }
            orderButton.text = prefs.order
            val adapter = DamagedShipAdapter(fragmentManager, prefs.sortType, prefs.order)
            recycler_view.swapAdapter(adapter, false)
            initDataList()
        }
        orderButton.text = this.prefs.order

        return rootView
    }


    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        recycler_view.layoutManager = LinearLayoutManager(activity)
        if (recycler_view.adapter == null) {
            val adapter = DamagedShipAdapter(fragmentManager, prefs.sortType, prefs.order)
            recycler_view.adapter = adapter
            initDataList()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        realm.close()
    }

    private fun initDataList() {
        (recycler_view.adapter as DamagedShipAdapter).clearData()
        val initialDataList = realm.where(MyShip::class.java).findAll()
                .filter { it.ndockTime > 0 }
        (recycler_view.adapter as DamagedShipAdapter).addDataOf(initialDataList)
        recycler_view.scrollToPosition(0)
    }

    companion object {

        fun newInstance(): DamagedShipFragment {
            return DamagedShipFragment()
        }
    }
}
